core/server/master/src/main/java/alluxio/master/file/ActiveSyncer.java

```diff
@@ -0,0 +1,104 @@
+package alluxio.master.file;
+
+import alluxio.AlluxioURI;
+import alluxio.SyncInfo;
+import alluxio.exception.InvalidPathException;
+import alluxio.heartbeat.HeartbeatExecutor;
+import alluxio.master.file.activesync.ActiveSyncManager;
+import alluxio.master.file.meta.MountTable;
+import alluxio.master.file.meta.options.MountInfo;
+import alluxio.resource.CloseableResource;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.io.PathUtils;
+import com.google.common.base.Throwables;
+import org.omg.CORBA.DynAnyPackage.Invalid;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Periodically sync the files for a particular mountpoint
+ */
+@NotThreadSafe
+public class ActiveSyncer implements HeartbeatExecutor {
+  private static final Logger LOG = LoggerFactory.getLogger(ActiveSyncer.class);
+  private final FileSystemMaster mFileSystemMaster;
+  private final ActiveSyncManager mSyncManager;
+  private final MountTable mMountTable;
+  private final String mRootPath;
+
+
+  /**
+   * Constructs a new {@link ActiveSyncer}.
+   */
+  public ActiveSyncer(FileSystemMaster fileSystemMaster, ActiveSyncManager syncManager,
+      MountTable mountTable, String rootPath) {
+    mFileSystemMaster = fileSystemMaster;
+    mSyncManager = syncManager;
+    mRootPath = rootPath;
+    mMountTable = mountTable;
+  }
+
+
+  @Override
+  public void heartbeat() {
+    LOG.info("start Active Syncer heartbeat");
+
+    List<AlluxioURI> filterList =  mSyncManager.getFilterList(mRootPath);
+    List<AlluxioURI> ufsUriList = filterList.stream().map(alluxioURI -> {
+      try {
+        return mMountTable.resolve(alluxioURI).getUri();
+      } catch (InvalidPathException e) {
+        LOG.warn("Invalid path " + alluxioURI.getPath());
+        return null;
+      }
+    }).collect(Collectors.toList());
+    if (filterList == null || filterList.isEmpty()) {
+      return;
+    }
+
+    LOG.info("filterList");
+    for (AlluxioURI filter: filterList) {
+      LOG.info("filterUri {}", filter.getPath());
+    }
+    for (AlluxioURI ufsUri: ufsUriList) {
+      LOG.info("ufsUri {}", ufsUri.getPath());
+    }
+
+    AlluxioURI path = filterList.get(0);
+    try {
+      MountTable.Resolution resolution = mMountTable.resolve(path);
+      AlluxioURI ufsUri = resolution.getUri();
+      try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
+        UnderFileSystem ufsClient = ufsResource.get();
+        if (ufsClient.supportsActiveSync()) {
```

Comment:
shouldn't this check be performed earlier, before the syncer heartbeat thread even starts?